### PR TITLE
Add test for add new client

### PR DESCRIPTION
**Pull request by beckmani**
Added new fixture: add_client
Added new test: Test that new client finishes bootstrap

### CODE REVIEW

File: tests/test_bs.py [hunk 1]
```diff
@@ -296,6 +328,34 @@ def fin():
     return _setup_clients_in_namespace(testconfig['namespace'])
 
 
+@pytest.fixture(scope='module')
+def add_client(request, setup_oracle, setup_poet, setup_bootstrap):
+
+    def _add_single_client():
+        bs_info = setup_bootstrap.pods[0]
```
Comment (antonlerner) [inline tests/test_bs.py:288]: if no bootstrap this will fail, add a message for that

File: tests/test_bs.py [hunk 2]
```diff
@@ -296,6 +328,34 @@ def fin():
...
+
+        client_args = {} if 'args' not in testconfig['client'] else testconfig['client']['args']
+
+        cspec = ContainerSpec(cname='client',
+                              cimage=testconfig['client']['image'],
+                              centry=[testconfig['client']['command']])
+
+        cspec.append_args(bootnodes="{0}:{1}/{2}".format(bs_info['pod_ip'], '7513', bs_info['key']),
```
Comment (antonlerner) [inline tests/test_bs.py]: move port to constant variable

File: tests/test_bs.py [hunk 3]
```diff
@@ -296,6 +328,34 @@ def fin():
...
+                          oracle_server='http://{0}:3030'.format(setup_oracle),
+                          poet_server='{0}:50002'.format(setup_poet),
```
Comment (antonlerner) [inline tests/test_bs.py]: same here

File: tests/test_bs.py [hunk 4]
```diff
@@ -392,7 +452,15 @@ def test_client(setup_clients, save_log_on_exit):
     peers = query_message(current_index, testconfig['namespace'], setup_clients.deployment_name, fields, True)
     assert peers == len(setup_clients.pods)
 
-    
+
+def test_add_client(add_client, setup_clients):
+    new_client = add_client()
+    time.sleep(20)
```
Comment (antonlerner) [inline tests/test_bs.py:414]: why is this sleep needed? the client has definitely booted

File: tests/test_bs.py [hunk 5]
```diff
@@ -67,6 +68,37 @@ def wait_to_deployment_to_be_ready(deployment_name, name_space, time_out=None):
...
+        # Set unique deployment id
+        if deployment_id:
+            dep['metadata']['generateName'] += '{0}-'.format(deployment_id)
+
+        if container_specs:
+            dep = container_specs.update_deployment(dep)
+
+        k8s_api = client.CoreV1Api()
+        resp = k8s_api.create_namespaced_pod(namespace=name_space, body=dep)
+        wait_for_pod_to_be_ready(resp.metadata._name, name_space, time_out=testconfig['deployment_ready_time_out'])
```
Comment (antonlerner) [inline tests/test_bs.py]: maybe we should have a different timeout for a single node than of a larger deployment

File: tests/k8s/single-client-w-conf.yml [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+metadata:
+  generateName: client-
+  labels:
+    run: client
+spec:
+  containers:
+  - name: client
+    image: spacemeshos/go-spacemesh:develop
+    imagePullPolicy: Always
+    args: ['--randcon', '3', '--bootstrap', '--test-mode', '--oracle_server', 'http://10.12.3.14:3030', '--grpc-server', '--json-server', '--bootnodes', '10.36.1.12:7513/26hRBJqMJPUnKgJ9VR6g3kH6aCQhZ9shBvhn3bEHDvRyv']
```
Comment (antonlerner) [inline tests/k8s/single-client-w-conf.yml:12]: shouldn't this be parameterised?

File: tests/test_bs.py (hunk 4 above)
Comment (beckmani) [reply tests/test_bs.py:414]: Added comment before

File: tests/test_bs.py (hunk 1 above)
Comment (beckmani) [reply tests/test_bs.py:288]: Done. Added check before.

File: tests/k8s/single-client-w-conf.yml (hunk 6 above)
Comment (beckmani) [reply tests/k8s/single-client-w-conf.yml:12]: It is. we override it later

**Review by antonlerner — APPROVED** [review]

File: tests/pod.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+def delete_pod(pod_name, name_space):
+    k8s_api = client.CoreV1Api()
+    try:
+        resp = k8s_api.delete_namespaced_pod(name=pod_name,
+                                             namespace=name_space,
+                                             body=client.V1DeleteOptions(propagation_policy='Foreground',
+                                                                         grace_period_seconds=5))
+    except ApiException as e:
+        if e.status == 404:
```
Comment (antonlerner) [inline tests/pod.py:65]: what about other errors?

File: tests/pod.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+            raise Exception("Timeout waiting to pod to be ready")
+
+
+def wait_to_pod_to_be_deleted(pod_name, name_space, time_out=None):
+    total_sleep_time = 0
+    while True:
+        try:
+            resp = client.CoreV1Api().read_namespaced_pod(name=pod_name, namespace=name_space)
+        except ApiException as e:
+            if e.status == 404:
```
Comment (antonlerner) [inline tests/pod.py:30]: what about other errors

File: tests/test_bs.py [hunk 9]
```diff
@@ -254,6 +261,54 @@ def fin():
...
+                          deployment_id=deployment_id,
+                          container_specs=container_specs)
+
+    client_name = resp.metadata.name
+    print("Add new client: {0}".format(client_name))
+    return client_name
+
+
+@pytest.fixture()
+def add_client(request, setup_oracle, setup_poet, setup_bootstrap, setup_clients):
```
Comment (antonlerner) [inline tests/test_bs.py:279]: add comment about this not being a functional fixture and shouldn't be used in our tests

File: tests/pod.py [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+        # Set unique deployment id
+        if deployment_id:
+            dep['metadata']['generateName'] += '{0}-'.format(deployment_id)
+
+        if container_specs:
+            dep = container_specs.update_deployment(dep)
+
+        k8s_api = client.CoreV1Api()
+        resp = k8s_api.create_namespaced_pod(namespace=name_space, body=dep)
```
Comment (antonlerner) [inline tests/pod.py:52]: if resp is error should we wait for pod to be ready?

File: tests/pod.py [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+from os import path
+from kubernetes import client
+from pytest_testconfig import config as testconfig
+from kubernetes.client.rest import ApiException
+
+
+def wait_for_pod_to_be_ready(pod_name, name_space, time_out=None):
+    total_sleep_time = 0
+    while True:
+        resp = client.CoreV1Api().read_namespaced_pod(name=pod_name, namespace=name_space)
```
Comment (antonlerner) [inline tests/pod.py:13]: can this return error? what should happen if so?

File: tests/pod.py [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+        if time_out and total_sleep_time > time_out:
+            raise Exception("Timeout waiting to pod to be ready")
+
+
+def wait_to_pod_to_be_deleted(pod_name, name_space, time_out=None):
+    total_sleep_time = 0
+    while True:
+        try:
+            resp = client.CoreV1Api().read_namespaced_pod(name=pod_name, namespace=name_space)
```
Comment (antonlerner) [inline tests/pod.py:28]: can this return error? what should happen if so?

File: tests/pod.py (hunk 7 above)
Comment (beckmani) [reply tests/pod.py:65]: This error may result from timing. (The pod may be delete before we get here)
Other error should not be ignored

File: tests/pod.py (hunk 11 above)
Comment (beckmani) [reply tests/pod.py:13]: I don't know. I havn't met an error here.
If we see that there are errors here we will handle them in the future

File: tests/test_bs.py (hunk 9 above)
Comment (beckmani) [reply tests/test_bs.py:279]: ok

**Review by antonlerner — APPROVED** [review]

**Pull request merged**